EIPS/eip-2003.md

```diff
@@ -0,0 +1,108 @@
+---
+eip: 2003
+title: EVMC modules for implementations of precompiled contracts
+author: Paweł Bylica (@chfast)
+discussions-to: https://github.com/ethereum/evmc/issues/259
+status: Draft
+type: Informational
+created: 2019-05-09
+requires: 1352
+---
+
+## Abstract
+
+[EVMC] specifies a generic API for Ethereum execution engines.
+This EIP specifies a way of providing implementations of Ethereum precompiled contracts
+using the [EVMC VM API].
+
+## Specification
+
+For the complete [EVMC] specification visit the [EVMC documentation first].
+
+The EVMC module with implementations of precompiled contracts SHOULD:
+
+1. Advertise the [`EVMC_CAPABILITY_PRECOMPILES`] capability 
+   in the [`get_capabilities()`] method.
+2. Implement the [`execute()`] method in the following way:
+   1. Check if the call destination address ([`evmc_message::destination`])
+      is within the range of precompiled contracts defined by [EIP-1352].
+      If not, abort execution with the [`EVMC_REJECTED`] status code.
+   2. Check if the call destination address ([`evmc_message::destination`])
+      targets existing precompiled contract.
+      Consider the EVM revision ([`evmc_revision`]) requested by 
+      the `rev` parameter of [`execute()`].
+      If yes, execute as follows:
+      1. Inspect the input data ([`evmc_message::input_data`], [`evmc_message::input_size`])
+         and calculate the _gas cost_ of the execution.
+      2. Compute the amount of _gas left_ after execution by
+         subtracting the _gas cost_ from the call gas limit ([`evmc_message::gas`]).
+      3. If _gas left_ is negative,
+         abort execution with the [`EVMC_OUT_OF_GAS`] status code.
+      4. Otherwise,
+         execute the code of the precompiled contract,
+         return the [`EVMC_SUCCESS`] status code, the output and _gas left_
+         ([`evmc_result::output_data`], [`evmc_result::output_size`], [`evmc_result::gas_left`]).
+   3. Otherwise, emulate execution of empty code by returning
```

Comment:
Not sure this emulation thing makes sense?

Reply:
Ping @chfast

Reply:
It does because then you have exclusive subsets of addresses - precompiles set and other, and the Precompiles VM is able to handle the matching set fully.

Reply:
This assumes that no code can be deployed in the precompiles address space.